### PR TITLE
i3: Default workspace & Workspace output assignment

### PR DESCRIPTION
**Pull request by SebTM**
### Description

Hello,

I'm a new to NixOS (and also nix, home-manager and the whole ecosystem). This is my first contribution after I tried to build my personal I3-config with home-manager and found some cases I was not able to do.

If there is a need I'm happy to adjust, please describe/comment in the code :-)
I tried my best to follow the contribution-guidelines and the checklist - if I missed something please inform me.

Cheers!

Possibly related to #695 but does not solve it completely.

### Checklist

- [X] Change is backwards compatible.

- [X] Code formatted with `./format`.

- [X] Code tested through `nix-shell --pure tests -A run.all`.

- [X] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [X] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

### CODE REVIEW

**Comment by sumnerevans** [comment]
Could you make this more generic so that it can be used in sway as well? I think that would involve moving more of the functions and the options to the `lib` folder.

File: modules/services/window-managers/i3-sway/i3.nix [hunk 1]
```diff
@@ -19,6 +19,13 @@ let
         workspaceAutoBackAndForth keycodebindings colors bars startup gaps menu
         terminal;
 
+      defaultWorkspace = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        description =
+          "Value should be sth. like 'workspace number X', ensures that this workspace definition goes first.";
```
Comment (rycee) [inline modules/services/window-managers/i3-sway/i3.nix:26]: "sth"?

File: modules/services/window-managers/i3-sway/i3.nix [hunk 2]
```diff
@@ -130,6 +137,19 @@ let
           for more advanced setup use 'i3.extraConfig'.
         '';
       };
+
+      variables = mkOption {
+        type = types.listOf (types.attrsOf (types.str));
+        default = [ ];
+        description =
+          "Ability to set variables at the beginning of the config e.g. for workspace-names.";
```
Comment (rycee) [inline modules/services/window-managers/i3-sway/i3.nix:145]:
```suggestion
        description = ''
          Ability to set variables at the beginning of the config, for
          example, for workspace-names.
        '';
```

File: modules/services/window-managers/i3-sway/i3.nix [hunk 3]
```diff
@@ -130,6 +137,19 @@ let
           for more advanced setup use 'i3.extraConfig'.
         '';
       };
+
+      variables = mkOption {
```
Comment (rycee) [inline modules/services/window-managers/i3-sway/i3.nix:141]: I don't know anything about i3 but this doesn't seem to make any sense. Why would you want to define variables as an option?

**Review by sumnerevans — CHANGES_REQUESTED** [review]
I think it would be good to split this into two PRs:

* add `defaultWorkspace` option.
* add `workspaceOutputAssign` option.

I agree with @rycee that it doesn't make much sense to have the `variables` option since that should be handled by Nix variables.

**Comment by rycee** [comment]
Please mind the commit message length, the subject line should at most 60 characters, [ideally 50](https://chris.beams.io/posts/git-commit/#limit-50). At the moment it seems to be 172 characters.

**Comment by SebTM** [comment]
Thanks for your Feedback!

- I moved the options to the lib-files
- I removed the variables option as it was really not needed, a "miss-thinking" in the first try I guess :)
- I adjusted the description to something more understandable
- I adjusted the commit message

I don't see the advantage of introducing a sortOrder to keybindings as there is no real impact on the rest of the orders as far as I found out? (and more complexity?)

If you give me feedback to just split up the PR, I'm happy to do this, until everything is ok it's easier for me just to keep up2date one :)

**Comment by sumnerevans** [comment]
Sorry for the delay getting back to this. I think it looks pretty good now. Let's go ahead and split it up into two PRs so that they are a bit easier to handle.

**Comment by SebTM** [comment]
No Problem, spitted into two PRs mentioned this :-)

**Pull request closed without merging**